apps/rush-lib/src/logic/taskRunner/Task.ts

```diff
@@ -6,55 +6,38 @@ import { ITaskWriter } from '@rushstack/stream-collator';
 import { Stopwatch } from '../../utilities/Stopwatch';
 import { TaskStatus } from './TaskStatus';
 import { TaskError } from './TaskError';
+import { BaseBuilder } from './BaseBuilder';
 
 /**
- * A definition for a task, an execute function returning a promise and a unique string name
+ * The `Task` class is a node in the dependency graph of work that needs to be scheduled by the `TaskRunner`.
+ * Each `Task` has a `BaseBuilder` member, whose subclass manages the actual operations for building a single
+ * project.
  */
-export interface ITaskDefinition {
+export class Task {
   /**
-   * Name of the task definition.
+   * When the scheduler is ready to process this `Task`, the `builder` implements the actual work of
+   * building the project.
    */
-  name: string;
+  public builder: BaseBuilder;
 
-  /**
-   * This flag determines if an incremental build is allowed for the task.
-   */
-  isIncrementalBuildAllowed: boolean;
-
-  /**
-   * Assigned by execute().  True if the build script was an empty string.  Operationally an empty string is
-   * like a shell command that succeeds instantly, but e.g. it would be odd to report build time statistics for it.
-   */
-  hadEmptyScript: boolean;
-
-  /**
-   * Method to be executed for the task.
-   */
-  execute: (writer: ITaskWriter) => Promise<TaskStatus>;
-}
-
-/**
- * The interface used internally by TaskRunner, which tracks the dependencies and execution status
- */
-export interface ITask extends ITaskDefinition {
   /**
    * The current execution status of a task. Tasks start in the 'ready' state,
    * but can be 'blocked' if an upstream task failed. It is 'executing' when
    * the task is executing. Once execution is complete, it is either 'success' or
    * 'failure'.
    */
-  status: TaskStatus;
+  public status: TaskStatus;
 
   /**
    * A set of all dependencies which must be executed before this task is complete.
    * When dependencies finish execution, they are removed from this list.
    */
-  dependencies: Set<ITask>;
+  public dependencies: Set<Task>;
 
   /**
    * The inverse of dependencies, lists all projects which are directly dependent on this one.
    */
-  dependents: Set<ITask>;
+  public dependents: Set<Task>;
```

Comment:
Make all of these `readonly` as well?

Reply:
Right now `Task` is just a container that `TaskRunner` writes stuff into. It needs to be redesigned to constrain when these properties get initialized. Again, out of scope for the problems I need to solve in this PR.